File: src/4node/apply/nodes/submitNode.ts
// ─── submitNode — 解析确认 + MCP 拉 userInfo + 提交申请 ─────────────────────
// 归属：apply 子图

import { HumanMessage, AIMessage } from '@langchain/core/messages'
import type { ApplyState } from '../../../3state/index.js'
import { getUserInfoTool, submitApplicationTool } from '../../../7controller/mcp/index.js'
import { parseCheckResults } from '../utils.js'

export async function submitNode(
  state: ApplyState,
  _config: { configurable: Record<string, unknown> }
): Promise<Partial<ApplyState>> {
  console.log('--apply:submit')

  const lastHuman = state.messages.filter(m => m instanceof HumanMessage).at(-1)
  const answer = String(lastHuman?.content ?? '').trim()

  let parsed: any
  try { parsed = JSON.parse(answer) }
  catch { parsed = { action: answer.toLowerCase() === 'cancel' ? 'cancel' : 'unknown' } }

  if (parsed.action === 'cancel') {
    return { messages: [new AIMessage('已取消申请，您可以随时重新发起。')] }
  }

  if (parsed.action !== 'confirm' || !Array.isArray(parsed.proofFileIds) || parsed.proofFileIds.length === 0) {
    return { messages: [new AIMessage('操作异常，请重试或联系管理员。')] }
  }

  const proofFileIds: number[] = parsed.proofFileIds
  const proofValues: number[]  = Array.isArray(parsed.proofValues) ? parsed.proofValues : []

  // 从 AsyncLocalStorage 自动取 userId，不再从 config.configurable 取
  const { getCurrentUserId } = await import('../../../7controller/mcp/index.js')
  const userId = getCurrentUserId()
  if (!userId) {
    return { messages: [new AIMessage('用户身份缺失，请重新登录后再申请。')] }
  }

  // MCP 调用 getUserInfo（无参数，userId 从 AsyncLocalStorage 取）
  const infoResult = await getUserInfoTool()
  if (!infoResult.success || !infoResult.data?.userInfo) {
    console.error('--apply:submit: MCP 拉取 userInfo 失败:', infoResult.error)
    return { messages: [new AIMessage('获取用户信息失败，请重新登录后再申请。')] }
  }
  const userInfo = infoResult.data.userInfo

  const suggestions = parseCheckResults(state.checkResults)
  const suggestion = suggestions[0]

  if (!suggestion) {
    return { messages: [new AIMessage('申请数据异常，请重新上传证明材料。')] }
  }

  const fullTemplate = state.templates.find(t => t.id === suggestion.templateId)

  // MCP 调用 submitApplication
  const submitResult = await submitApplicationTool({
    templateName: suggestion.templateName,
    applyScore: suggestion.estimatedScore,
    ruleId: suggestion.ruleId ?? undefined,
    remark: `AI 智能匹配 - ${suggestion.reason ?? ''}`,
    proofItems: proofFileIds.map((id: number, i: number) => ({
      proofFileId: id,
      proofValue: proofValues[i] ?? 0,
    })),
  })

  if (submitResult.success) {
    const appId = (submitResult.data as any)?.applicationId ?? '未知'
    return { messages: [new AIMessage(
      `✅ 申请已提交成功！申请编号：**${appId}**，请等待审核员审核。`
    )] }
  }

  return { messages: [new AIMessage(`提交失败：${submitResult.error}，请稍后重试或手动提交。`)] }
}

File: src/nodes/apply/submit_node.py
"""submitNode - 提交节点 - 对应 TS: submitNode.ts"""
import json
from typing import Any
from langchain_core.messages import HumanMessage, AIMessage
from langchain_core.messages import BaseMessage
from tools.backend_client import BackendClient


async def submit_node(state: dict, config: dict) -> dict:
    """
    提交节点
    
    归属: apply 子图
    功能: 将申请提交到后端
    """
    print("--apply:submit")
    
    check_results: list[str] = state.get("check_results", [])
    messages: list[BaseMessage] = state.get("messages", [])
    
    # 从 config 获取参数
    user_token = config.get("configurable", {}).get("user_token", "")
    thread_id = config.get("configurable", {}).get("thread_id", "")
    
    # 解析 suggestions
    suggestions = []
    for r in check_results:
        try:
            parsed = json.loads(r)
            if "error" not in parsed:
                suggestions.append({
                    "templateId": parsed.get("templateId"),
                    "ruleId": parsed.get("ruleId"),
                    "estimatedScore": parsed.get("estimatedScore"),
                })
        except json.JSONDecodeError:
            continue
    
    if not suggestions:
        return {
            "messages": [AIMessage(content="提交失败：没有有效的加分项")]
        }
    
    # 提交到后端
    result = await BackendClient.submit_application(
        user_token=user_token,
        session_id=thread_id,
        suggestions=suggestions,
    )
    
    if result.get("success"):
        reply = "加分申请已提交成功！请等待审核。"
    else:
        reply = f"提交失败：{result.get('error', '未知错误')}"
    
    return {
        "messages": [AIMessage(content=reply)]
    }
